Realm/Realm/Sync/ErrorHandling/DiscardLocalResetHandler.cs

```diff
@@ -0,0 +1,67 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+namespace Realms.Sync.ErrorHandling
+{
+    /// <summary>
+    /// A client reset strategy where all the not yet synchronized data is automatically discarded and a fresh copy of the synchronized Realm is obtained.
+    /// To be noted that the freshly downloaded copy of the synchronized Realm triggers all change notifications as a write transaction is internally simulated.
+    /// This strategy supplies three callbacks: <see cref="OnBeforeReset"/> and <see cref="OnAfterReset"/> to give a chance to take some user actions,
+    /// respectively, before and after a client reset has happened.
+    /// And in case something goes wrong during the client reset, <see cref="ManualResetFallback"/> is supplied to take specific actions in such condition.
+    /// For more information see <see cref="ClientResetHandlerBase"/>.
+    /// </summary>
+    /// <seealso href="https://docs.mongodb.com/realm/sync/overview/">Sync Overview Docs</seealso>
+    public sealed class DiscardLocalResetHandler : ClientResetHandlerBase
+    {
+        /// <summary>
+        /// Callback that indicates a Client Reset is about to happen.
+        /// </summary>
+        /// <param name="beforeFrozen">
+        /// The frozen <see cref="Realm"/> before the reset.
+        /// </param>
+        public delegate void BeforeResetCallback(Realm beforeFrozen);
+
+        /// <summary>
+        /// Callback that indicates a Client Reset has just happened.
+        /// </summary>
+        /// <param name="beforeFrozen">
+        /// The frozen <see cref="Realm"/> as it was before the reset.
+        /// </param>
+        /// <param name="after">
+        /// The <see cref="Realm"/> after the client reset.
+        /// </param>
+        public delegate void AfterResetCallback(Realm beforeFrozen, Realm after);
```

Comment:
For these callbacks, we should clarify whether `after` is already in a write transaction or one should create their own as well as that the lifetime of these Realms and all objects within them is scoped to the callback - i.e. you can't store references to objects or the Realm itself in `Before` and expect them to be valid in `After`.

Reply:
I'm not sure I fully understand the lifetime problem.`beforeFrozen` is frozen, so refs to realm objects are not a problem as long as the memory growth, given by version pinning, is acceptable to the use case.
Saving a ref to a realm in `OnBefore` to use it in `OnAfter`, as you pointed out already, makes little sense as `OnAfter` gives you a ref to both the before and the after realm.
Are you specifically asking about a **non** realm object created in `OnBefore` to be used in `OnAfter`? I'm not really sure here.